torch/csrc/jit/passes/specialize_autogradzero.cpp

```diff
@@ -16,7 +16,8 @@ void specializeAutogradZero(Graph& g) {
 
   for (Value* input : g.inputs()) {
     const auto& tp = input->type();
-    if (tp->isSubtypeOf(AutogradZeroTensorType::get())) {
+    if (tp->isSubtypeOf(AutogradZeroTensorType::get()) ||
+        tp->isSubtypeOf(NoneType::get())) {
```

Comment:
This is unnecessary. We never use `Optional[Tensor]` for those inputs, so it will never get triggered. It might actually be a good idea to get rid of `AutogradZeroTensorType` though (in favor of optionals).

Reply:
Well. I get failures without it. :confused:

Reply:
What kinds of failures? I would really like us to understand what's happening instead of applying patches blindly

Reply:
So what happens is for Optional tensor arguments and None inputs, the `ArgumentInfo` (`argument_spec.h`) will have `defined=false`. If now the backward graph takes an `OptionalType` argument, we specialize to `None` (instead of `AutogradZeroTensorType`).
But this then confuses the above if-statement, which differentiates between not-passed-in Tensors and passed-in-Tensors, until now by using `AutogradZeroTensorType`.
In a way, we're involuntarily taking a tiny step towards getting rid of `AutogradZeroTensorType` (which would involve making the replacement in the `ArgumentSpec` unconditional and then one could probably remove the test for `AutogradZeroTensorType` above).

Reply:
This is gone now with better None vs. undefined handling.